torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -533,34 +533,38 @@ class FunctionInliner : public IRMutator {
         outputs_(std::move(outputs)) {
     for (auto* i : producer->indices()) {
       const Var* index_var = dynamic_cast<const Var*>(i);
-      if (index_var == nullptr) {
+      if (dynamic_cast<const Var*>(i) != nullptr) {
+        index_vars_.insert(index_var);
+        producer_index_vars_.push_back(index_var);
+      } else if (dynamic_cast<const IntImm*>(i) != nullptr) {
+        // If the index can be a constant, then that dimension must have size 1
+        // (since we don't support in-place writes). Resolves issue 52581.
+        TORCH_INTERNAL_ASSERT(
+            dynamic_cast<const IntImm*>(i)->value() == 0,
+            "It shouldn't be possbile for this to ever be non-zero");
+        producer_index_vars_.push_back(nullptr);
+      } else {
         throw std::logic_error("cannot inline Buf with compound indices");
       }
-      index_vars_.insert(index_var);
-      producer_index_vars_.push_back(index_var);
     }
   }
 
  private:
-  // For the target function, insert the caller/callee pair into the replacement
-  // mapping.
-  const Expr* mutate(const FunctionCall* v) override {
-    const Tensor* t = v->tensor();
-    const Buf* buf = t->buf();
-    if (buf != buf_) {
-      return IRMutator::mutate(v);
-    }
-
-    if (v->nparams() != buf->ndim()) {
-      throw malformed_input(
-          "Placeholder indexed access is inconsistent with its rank", v);
-    }
-
+  const Expr* mutate_loads(const Buf* buf, std::vector<const Expr*> dims) {
```

Comment:
Kind of a tip to make future diffs like this easier to review, it's nice to stack changes that include refactoring and functionality changes as two diffs: (1) refactor, (2) new behavior.  That way I can glance through (1) to verify it's just code motion, and then use my actual brainpower (such as it is) on the functional changes.

If you use ghstack it makes the stacked diff experience basically as nice as phabricator :-)

Reply:
right, I'll switch to ghstack this weekend :P